stackslib/src/net/api/postblock_proposal.rs

```diff
@@ -0,0 +1,489 @@
+                reason_code: ValidateRejectCode::BadBlockHash,
+            });
+        }
+
+        info!(
+            "Participant: validated anchored block";
+            "block_header_hash" => %computed_block_header_hash,
+            "height" => block.header.chain_length,
+            "tx_count" => block.txs.len(),
+            "parent_stacks_block_id" => %block.header.parent_block_id,
+            "block_size" => size,
+            "execution_cost" => %cost,
+            "validation_time_ms" => time_elapsed(),
+            "tx_fees_microstacks" => block.txs.iter().fold(0, |agg: u64, tx| {
+                agg.saturating_add(tx.get_tx_fee())
+            })
+        );
+
+        Ok(BlockValidateOk { block, cost, size })
+    }
+}
+
+#[derive(Clone, Default)]
+pub struct RPCBlockProposalRequestHandler {
+    pub block_proposal: Option<NakamotoBlockProposal>,
+}
+
+impl RPCBlockProposalRequestHandler {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Decode a JSON-encoded block proposal
+    fn parse_json(body: &[u8]) -> Result<NakamotoBlockProposal, Error> {
+        serde_json::from_slice(body)
+            .map_err(|e| Error::DecodeError(format!("Failed to parse body: {e}")))
+    }
+}
+
+/// Decode the HTTP request
+impl HttpRequest for RPCBlockProposalRequestHandler {
+    fn verb(&self) -> &'static str {
+        "POST"
+    }
+
+    fn path_regex(&self) -> Regex {
+        Regex::new(r#"^/v2/block_proposal$"#).unwrap()
+    }
+
+    /// Try to decode this request.
+    /// There's nothing to load here, so just make sure the request is well-formed.
+    fn try_parse_request(
+        &mut self,
+        preamble: &HttpRequestPreamble,
+        _captures: &Captures,
+        query: Option<&str>,
+        body: &[u8],
+    ) -> Result<HttpRequestContents, Error> {
+        // Only accept requests from localhost
+        let is_loopback = match preamble.host {
+            // Should never be DNS
+            PeerHost::DNS(..) => false,
+            PeerHost::IP(addr, ..) => addr.is_loopback(),
+        };
+
+        if !is_loopback {
+            return Err(Error::Http(403, "Forbidden".into()));
+        }
+
+        if preamble.get_content_length() == 0 {
+            return Err(Error::DecodeError(
+                "Invalid Http request: expected non-zero-length body for block proposal endpoint"
+                    .to_string(),
+            ));
+        }
+
+        if preamble.get_content_length() > MAX_PAYLOAD_LEN {
+            return Err(Error::DecodeError(
+                "Invalid Http request: BlockProposal body is too big".to_string(),
+            ));
+        }
+
+        let block_proposal = match preamble.content_type {
+            Some(HttpContentType::JSON) => Self::parse_json(body)?,
+            None => {
+                return Err(Error::DecodeError(
+                    "Missing Content-Type for block proposal".to_string(),
+                ))
+            }
+            _ => {
```

Comment:
This is just a side note, but I had a brain fart right here reading this. Would be clearer if you reorged it to 

```
...
            Some(HttpContentType::JSON) => Self::parse_json(body)?,
            Some(_) => {
                return Err(Error::DecodeError(
                    "Wrong Content-Type for block proposal; expected application/json".to_string(),
                ))
            }
            None => {
                return Err(Error::DecodeError(
                    "Missing Content-Type for block proposal".to_string(),
                ))
            }